### PR TITLE
feat(source): support extract pk for debezium avro source

### PR DESCRIPTION
**Pull request by waruto210**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- support extract pk for debezium avro source

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).~
~- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Connector (sources & sinks)

### Release note

</details>

User can create a table with debezium avro source without specify pk columns wihch would be extract from kafka key schema.

```SQL
create table t with (
    connector = 'kafka',
    topic = 'dbserver1.inventory.customers',
    properties.bootstrap.server = '127.0.0.1:9092',
    scan.startup.mode = 'earliest')
row format debezium_avro
row schema location confluent schema registry 'http://127.0.0.1:8081';
```

### CODE REVIEW

File: src/connector/src/parser/debezium/avro_parser.rs [hunk 1]
```diff
@@ -66,10 +69,17 @@ impl DebeziumAvroParserConfig {
         })?;
         let kafka_topic = get_kafka_topic(props)?;
         let client = Client::new(url, props)?;
+        let cf_schema = client
```
[the remark this answers is not in the document]
Comment (waruto210) [reply src/connector/src/parser/debezium/avro_parser.rs:72]: > What does it mean by "cf"

Schema from confluent, it contains an id and a raw string of real schema.

[the remark this answers is not in the document]
Comment (waruto210) [reply src/connector/src/parser/debezium/avro_parser.rs:72]: Without pk, we can't do update/delete.

[the remark this answers is not in the document]
Comment (waruto210) [reply src/connector/src/parser/debezium/avro_parser.rs:72]: > > Without pk, we can't do update/delete.
> 
> I mean when no PK is specified from the MySQL side. Will the program get wrong?

Our code will get wrong, however, in this case, debezium can only send insert records, and this scenario is very specific.

File: src/frontend/src/handler/create_source.rs [hunk 2]
```diff
@@ -119,8 +137,8 @@ pub(crate) async fn resolve_source_schema(
     source_schema: SourceSchema,
     columns: &mut Vec<ColumnCatalog>,
     with_properties: &HashMap<String, String>,
-    row_id_index: Option<usize>,
-    pk_column_ids: &[ColumnId],
+    row_id_index: &mut Option<usize>,
+    pk_column_ids: &mut Vec<ColumnId>,
```
[the remark this answers is not in the document]
Comment (waruto210) [reply src/frontend/src/handler/create_source.rs:141]: Yes, user-defined schema is not allowed with row format debezium_avro, but `pk_column_ids` will never be empty, if user do not specify a pk, `pk_column_ids` will contain `ROW_ID_COLUMN_ID`.

**Comment by hzxa21** [comment]
In some use cases, key schema may be absent in scheme registry or the key may not be in avro format even though dbz is used. Therefore, we should provide the flexibility to user to configure the key format/schema. There are two approaches:
1. Allow user to define the pk fields in `CREATE TABLE`, just like what we did before this PR. If pk fields is provided, we will skip looking at the kakfa message key.
2. Allow user to define the `KEY FORMAT` and allow it to be different from the `ROW FORMAT`.

I prefer 1 > 2 since it is simpler to implement and easier to understand for user given that we haven't seen a case that the pk field is missing in the kafka message value.

**Comment by waruto210** [comment]
> 

I also prefer 1 > 2.

**Comment by waruto210** [comment]
> I prefer 1 > 2 since it is simpler to implement and easier to understand for user given that we haven't seen a case that the pk field is missing in the kafka message value.

I also prefer 1 > 2.

File: src/connector/src/parser/debezium/avro_parser.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (waruto210) [reply src/connector/src/parser/debezium/avro_parser.rs]: > I mean when no PK is specified from the MySQL side. Will the program get wrong?

After discussing with @hzxa21, we believe that for now we should have to ask the user to specify pk(s), or the schema-registry has a schema of pk.

Our decision is based on the following reason:

If a MySQL table doesn't have pk(s), MySQl will add a record_id as it's pk, but the debezium connector will emit c/u/d events without the record_id, we can't process such events.

[the remark this answers is not in the document]
Comment (tabVersion) [reply src/connector/src/parser/debezium/avro_parser.rs]: we may do an example to see if debezium prunes record_id column.

File: src/connector/src/parser/debezium/avro_parser.rs [hunk 3]
```diff
@@ -116,6 +126,20 @@ impl DebeziumAvroParserConfig {
...
+    pub fn get_pk_names(&self) -> Result<Vec<String>> {
+        Self::get_pk_names_inner(&self.key_schema)
+    }
+
+    pub(crate) fn get_pk_names_inner(key_schema: &Schema) -> Result<Vec<String>> {
+        if let Schema::Record { fields, .. } = key_schema {
+            Ok(fields.iter().map(|field| field.name.clone()).collect_vec())
+        } else {
+            Err(RwError::from(InternalError(
+                "inner avro key schema invalid, record required".into(),
```
Comment (tabVersion) [inline src/connector/src/parser/debezium/avro_parser.rs:138]: the error looks strange to me, may refer to the top level message should be a `record`

File: src/connector/src/parser/debezium/avro_parser.rs [hunk 4]
```diff
@@ -132,7 +156,7 @@ impl DebeziumAvroParserConfig {
             Ok(fields)
         } else {
             Err(RwError::from(InternalError(
-                "inner avro schema invalid, record required".into(),
+                "inner avro value schema invalid, record required".into(),
```
Comment (tabVersion) [inline src/connector/src/parser/debezium/avro_parser.rs:159]: ditto

File: src/frontend/src/handler/create_source.rs [hunk 5]
```diff
@@ -268,39 +279,70 @@ pub(crate) async fn resolve_source_schema(
...
-                    })?;
-                let _ = full_columns.remove(index);
+            // no row_id, so user specify pk(s)
+            if row_id_index.is_none() {
+                // user specify no only pk columns
+                if columns.len() != pk_column_ids.len() {
+                    return Err(RwError::from(ProtocolError(
+                        "User can only specify primary key columns when creating table with row
+                        format debezium_avro."
+                            .to_owned(),
```
Comment (tabVersion) [inline src/frontend/src/handler/create_source.rs:289]: so 
```sql
create source s(xxx int, yyy int, ...) with ( ... ) row format debezium_avro schema registry http://xxx
```
is not allowed?

what if user specified columns are not consistent with columns in schema registry

Comment (waruto210) [reply src/frontend/src/handler/create_source.rs:289]: > so
> 
> ```sql
> create source s(xxx int, yyy int, ...) with ( ... ) row format debezium_avro schema registry http://xxx
> ```
> 
> is not allowed?
> 

Yes, user can only specify pk column(s) or nothing.

> what if user specified columns are not consistent with columns in schema registry

We use the pk column names specified by user to find real pk_column_ids in schema extracted from schema registry, if they do not match, the program will error.

**Review by tabVersion — APPROVED** [review]
LGTM
good job!

**Pull request merged**